File: src/main/java/org/kite9/diagram/visualization/compaction/segment/SegmentBuilder.java
package org.kite9.diagram.visualization.compaction.segment;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import org.kite9.diagram.common.algorithms.det.UnorderedSet;
import org.kite9.diagram.common.algorithms.so.AlignStyle;
import org.kite9.diagram.common.elements.Dimension;
import org.kite9.diagram.common.elements.edge.Edge;
import org.kite9.diagram.common.elements.vertex.FanVertex;
import org.kite9.diagram.common.elements.vertex.Vertex;
import org.kite9.diagram.model.AlignedRectangular;
import org.kite9.diagram.model.Connection;
import org.kite9.diagram.model.Container;
import org.kite9.diagram.model.Rectangular;
import org.kite9.diagram.model.position.Direction;
import org.kite9.diagram.model.style.DiagramElementSizing;
import org.kite9.diagram.model.style.HorizontalAlignment;
import org.kite9.diagram.model.style.VerticalAlignment;
import org.kite9.diagram.visualization.orthogonalization.Dart;
import org.kite9.diagram.visualization.orthogonalization.Orthogonalization;
import org.kite9.framework.common.Kite9ProcessingException;
import org.kite9.framework.logging.Kite9Log;
import org.kite9.framework.logging.Logable;
import org.kite9.framework.logging.LogicException;


/**
 * This looks at the orthogonal representation and works out from the available Darts what 
 * Vertices must be on the same Vertical or Horizontal line.
 * 
 * @author robmoffat
 *
 */
public class SegmentBuilder implements Logable {
	
	Kite9Log log = new Kite9Log(this);
	
	public List<Segment> buildSegmentList(Orthogonalization o, Set<Direction> planeDirection, Dimension direction) {
		Set<Direction> transversePlane = new LinkedHashSet<Direction>();
		for (Direction d2 : planeDirection) {
			transversePlane.add(Direction.rotateClockwise(d2));
		}
		
		List<Segment> result = new ArrayList<Segment>();
		Set<Vertex> done = new UnorderedSet<Vertex>();
		int segNo = 1;
	
		for (Dart d : o.getAllDarts()) {
			if (planeDirection.contains(d.getDrawDirection())) {
				Vertex v = d.getFrom();
				if (!done.contains(v)) {
					Segment s = new Segment(direction, segNo++);
					extendSegmentFromVertex(v, planeDirection, s, done);
					s.setAlignStyle(getSegmentAlignStyle(s));
					done.addAll(s.getVerticesInSegment());
					result.add(s);
				}
			}
		}
		
		log.send("Segments", result);
		
		return result;
	}
	

	public AlignStyle getSegmentAlignStyle(Segment s) {
		Set<Connection> conns = s.getConnections();
		if (conns.size() == 1) {
			Connection de = conns.iterator().next();
			return decideConnectionSegmentAlignStyle(s, de);
		} else if (conns.size() == 0) {
			UnderlyingInfo toUse = s.getUnderlyingInfo().stream()
				.filter(ui -> ui.getDiagramElement() instanceof AlignedRectangular)
				.sorted((a, b) -> ((Integer) ((Rectangular) a.getDiagramElement()).getDepth())
					.compareTo(((Rectangular)b.getDiagramElement()).getDepth()))
				.findFirst().orElse(null);
			
			
			if (toUse != null) {
				AlignStyle out = decideRectangularAlignStyle(s, (AlignedRectangular) toUse.getDiagramElement()); 
				return out;
			}
			
			return null;
			
		} else {
			throw new LogicException();
		}
		
	}
	
	private AlignStyle decideRectangularAlignStyle(Segment s, AlignedRectangular de) {
		DiagramElementSizing des = de instanceof Container ? ((Container)de).getSizing() : null;
		
		if ((des == DiagramElementSizing.MINIMIZE) || (des == null)) {
			if (s.getDimension() == Dimension.H) {
				VerticalAlignment va = de.getVerticalAlignment();
				switch (va) {
				case BOTTOM:
					return AlignStyle.MAX;
				case CENTER:
					return AlignStyle.CENTER;
				case TOP:
					return AlignStyle.MIN;
				}
			} else if (s.getDimension() == Dimension.V) {
				HorizontalAlignment ha = de.getHorizontalAlignment();
				switch (ha) {
				case LEFT:
					return AlignStyle.MIN;
				case CENTER:
					return AlignStyle.CENTER;
				case RIGHT:
					return AlignStyle.MAX;
				}
			}
		}
		
		return null;
	}

	private AlignStyle decideConnectionSegmentAlignStyle(Segment s, Connection de) {
		if (de.getRenderingInformation().isContradicting()) {
			return null;
		}
		
		if (s.getDimension() == Dimension.H) {
			// horizontal segment, push up or down
			Set<Direction> pushDirections = filterFanDirections(s, Direction::isVertical);
			
//			if (pushDirections.size() > 1) {
//				throw new Kite9ProcessingException();
//			}
			
			if (pushDirections.size() == 1) {
				for (Direction d : pushDirections) {
					switch (d) {
					case UP:
						return AlignStyle.MIN;
					case DOWN:
						return AlignStyle.MAX;
					default:
					}
				}
			}
			
			return AlignStyle.CENTER;
		} else if (s.getDimension() == Dimension.V) {
			Set<Direction> pushDirections = filterFanDirections(s, Direction::isHorizontal);
			
//			if (pushDirections.size() > 1) {
//				throw new Kite9ProcessingException();
//			}
			
			if (pushDirections.size() == 1) {
				for (Direction d : pushDirections) {
					switch (d) {
					case LEFT:
						return AlignStyle.MIN;
					case RIGHT:
						return AlignStyle.MAX;
					default:
					}
				}
			}
			
			return AlignStyle.CENTER;
		} else {
			throw new Kite9ProcessingException("No dimension on segment");
		}
	}


	private Set<Direction> filterFanDirections(Segment s, Predicate<? super Direction> axis) {
		return s.getVerticesInSegment().stream()
			.filter(v -> v instanceof FanVertex)
			.map(v -> (FanVertex)v)
			.flatMap(v -> v.getFanSides().stream())
			.filter(axis)
			.collect(Collectors.toSet());
	}


	private void extendSegmentFromVertex(Vertex v, Set<Direction> planeDirection, Segment samePlane, Set<Vertex> done) {
		if (done.contains(v)) 
			return;
		
		samePlane.addToSegment(v);
		done.add(v);
		for (Edge e : v.getEdges()) {
			if ((e instanceof Dart) && (planeDirection.contains(((Dart)e).getDrawDirection()))) {
				Vertex other = e.otherEnd(v);
				extendSegmentFromVertex(other, planeDirection, samePlane, done);
			}
		}
	}

	public String getPrefix() {
		return "SEGB";
	}

	public boolean isLoggingEnabled() {
		return true;
	}
	
}

File: k9layout/src/main/kotlin/org/kite9/diagram/visualization/compaction/segment/SegmentBuilder.kt
package org.kite9.diagram.visualization.compaction.segment

import org.kite9.diagram.common.algorithms.det.UnorderedSet
import org.kite9.diagram.common.algorithms.so.AlignStyle
import org.kite9.diagram.common.elements.Dimension
import org.kite9.diagram.common.elements.vertex.FanVertex
import org.kite9.diagram.common.elements.vertex.Vertex
import org.kite9.diagram.logging.Kite9Log
import org.kite9.diagram.logging.Logable
import org.kite9.diagram.logging.LogicException
import org.kite9.diagram.model.AlignedRectangular
import org.kite9.diagram.model.Connection
import org.kite9.diagram.model.SizedRectangular
import org.kite9.diagram.model.position.Direction
import org.kite9.diagram.model.position.Direction.Companion.rotateClockwise
import org.kite9.diagram.model.style.DiagramElementSizing
import org.kite9.diagram.model.style.HorizontalAlignment
import org.kite9.diagram.model.style.VerticalAlignment
import org.kite9.diagram.visualization.orthogonalization.Dart
import org.kite9.diagram.visualization.orthogonalization.Orthogonalization

/**
 * This looks at the orthogonal representation and works out from the available Darts what
 * Vertices must be on the same Vertical or Horizontal line.
 *
 * @author robmoffat
 */
class SegmentBuilder : Logable {

    var log = Kite9Log.instance(this)

    fun buildSegmentList(o: Orthogonalization, planeDirection: Set<Direction>, direction: Dimension): List<Segment> {
        val transversePlane: MutableSet<Direction> = LinkedHashSet()
        for (d2 in planeDirection) {
            transversePlane.add(rotateClockwise(d2))
        }
        val result: MutableList<Segment> = ArrayList()
        val done: MutableSet<Vertex> = UnorderedSet()
        var segNo = 1
        for (d in o.getAllDarts()) {
            if (planeDirection.contains(d.getDrawDirection())) {
                val v = d.getFrom()
                if (!done.contains(v)) {
                    val s = Segment(
                        direction, segNo++
                    )
                    extendSegmentFromVertex(v, planeDirection, s, done)
                    s.alignStyle = getSegmentAlignStyle(s)
                    done.addAll(s.getVerticesInSegment())
                    result.add(s)
                }
            }
        }
        log.send("Segments", result)
        return result
    }

    fun getSegmentAlignStyle(s: Segment): AlignStyle? {
        val conns = s.connections
        return if (conns.size == 1) {
            val de = conns.iterator().next()
            decideConnectionSegmentAlignStyle(s, de)
        } else if (conns.size == 0) {
            val toUse : AlignedRectangular? = s.underlyingInfo
                .map { it.diagramElement }
                .filterIsInstance<AlignedRectangular>()
                .sortedWith  { a : AlignedRectangular, b: AlignedRectangular -> a.getDepth().compareTo(b.getDepth()) }
                .firstOrNull()
            if (toUse != null) {
                return decideRectangularAlignStyle(s, toUse)
            }
            null
        } else {
            throw LogicException()
        }
    }

    private fun decideRectangularAlignStyle(s: Segment, de: AlignedRectangular): AlignStyle? {
        val des = if (de is SizedRectangular) (de as SizedRectangular).getSizing(s.dimension === Dimension.H) else null
        if (des === DiagramElementSizing.MINIMIZE || des == null) {
            if (s.dimension === Dimension.H) {
                val va = de.getVerticalAlignment()
                return when (va) {
                    VerticalAlignment.BOTTOM -> AlignStyle.MAX
                    VerticalAlignment.CENTER -> AlignStyle.CENTER
                    VerticalAlignment.TOP -> AlignStyle.MIN
                }
            } else if (s.dimension === Dimension.V) {
                val ha = de.getHorizontalAlignment()
                return when (ha) {
                    HorizontalAlignment.LEFT -> AlignStyle.MIN
                    HorizontalAlignment.CENTER -> AlignStyle.CENTER
                    HorizontalAlignment.RIGHT -> AlignStyle.MAX
                }
            }
        }
        return null
    }

    private fun decideConnectionSegmentAlignStyle(s: Segment, de: Connection): AlignStyle? {
        if (de.getRenderingInformation().isContradicting) {
            return null
        }
        return if (s.dimension === Dimension.H) {
            // horizontal segment, push up or down
            val pushDirections = filterFanDirections(s) { Direction.isVertical( it ) }

//			if (pushDirections.size() > 1) {
//				throw new Kite9ProcessingException();
//			}
            if (pushDirections.size == 1) {
                for (d in pushDirections) {
                    when (d) {
                        Direction.UP -> return AlignStyle.MIN
                        Direction.DOWN -> return AlignStyle.MAX
                        else -> {
                        }
                    }
                }
            }
            AlignStyle.CENTER
        } else if (s.dimension === Dimension.V) {
            val pushDirections = filterFanDirections(s) { Direction.isHorizontal(it) }

//			if (pushDirections.size() > 1) {
//				throw new Kite9ProcessingException();
//			}
            if (pushDirections.size == 1) {
                for (d in pushDirections) {
                    when (d) {
                        Direction.LEFT -> return AlignStyle.MIN
                        Direction.RIGHT -> return AlignStyle.MAX
                        else -> {
                        }
                    }
                }
            }
            AlignStyle.CENTER
        } else {
            throw LogicException("No dimension on segment")
        }
    }

    private fun filterFanDirections(s: Segment, axis: (Direction) -> Boolean): Set<Direction> {
        return s.getVerticesInSegment()
            .filterIsInstance<FanVertex>()
            .flatMap { it.fanSides }
            .filter { axis (it) }
            .toSet()
    }

    private fun extendSegmentFromVertex(
        v: Vertex,
        planeDirection: Set<Direction?>,
        samePlane: Segment,
        done: MutableSet<Vertex>
    ) {
        if (done.contains(v)) return
        samePlane.addToSegment(v)
        done.add(v)
        for (e in v.getEdges()) {
            if (e is Dart && planeDirection.contains(e.getDrawDirection())) {
                val other = e.otherEnd(v)
                extendSegmentFromVertex(other, planeDirection, samePlane, done)
            }
        }
    }

    override val prefix: String
        get() = "SEGB"

    override val isLoggingEnabled: Boolean
        get() = true
}